File: 백준/Silver/1914. 하노이 탑/하노이 탑.py
import sys
input = sys.stdin.readline

def hanoi(n, A, B, C):  # 시작, 종료, 경유지
    global cnt
    if n == 1:
        print(A, B)
        return
    hanoi(n - 1, A, C, B)   # 우선 경유지에 n-1개를 가져다 놓아야함
    print(A, B)         # 시작점에서 도착점에 가장 큰 것을 하나두고
    hanoi(n - 1, C, B, A)   # 경유지에 있던 n-1개의 원판을 도착점으로


N = int(input())
dp = [0] * (N + 1)
dp[1] = 1
for i in range(2, N + 1):
    dp[i] = 2 * dp[i - 1] + 1
print(dp[N])
if N <= 20:
    hanoi(N, 1, 3, 2)

File: 백준/Silver/1914. 하노이 탑/하노이 탑.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.math.BigInteger;
import java.util.StringTokenizer;

public class Main {
    public void Hanoi(StringBuilder sb,  int N, int start, int mid, int des){ // N개가 start, mid, des 위치로 들어온다.
        if(N == 0) return;
        Hanoi(sb, N - 1, start, des, mid); // 가장 아래를 빼고 N - 1개는 위쪽 des로 가기 위해서 목적지를 가운데로 한다
        sb.append(start + " " + des + "\n"); // 가장 아래 원판을 des로 옮긴다
        Hanoi(sb,  N - 1, mid, start, des); // N - 1개가 가운데로 왔으니 des로 옮겨야 한다.
    }

    public static void main(String[] args) throws IOException {
        Main T = new Main();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        BigInteger minus = new BigInteger("1");
        BigInteger cnt = new BigInteger("2").pow(N).subtract(minus);

        if(N <= 20) {
            T.Hanoi(sb,  N, 1, 2, 3);
            System.out.println(cnt);
            System.out.println(sb);
        } else {
            System.out.println(cnt);
        }

    }
}
